theano/tensor/nnet/nnet.py

```diff
@@ -786,16 +934,102 @@ def local_logsoftmax_grad(node):
         copy_stack_trace(node.outputs[0], ret)
         return [ret]
 
 
+@opt.register_specialize('fast_compile')
+@gof.local_optimizer([SoftmaxGrad])
+def local_indexing_logsoftmax_grad(node):
+    # One case is supported:
+    # 1. AdvancedIncSubtensor(
+    #           zeros_like(softmax(x)),
+    #           -out_grad / AdvancedSubtensor(softmax(x), arange(y.shape[0]), y),
+    #           arange(y.shape[0]),
+    #           y)
+    #   which arises from the gradient of log(softmax(x)[arange(y.shape[0]), y])
+    list_classes_IncSubtensor = ((IncSubtensor, AdvancedIncSubtensor1, AdvancedIncSubtensor))
+    list_classes_Subtensor = ((Subtensor, AdvancedSubtensor1, AdvancedSubtensor))
+    if not (isinstance(node.op, SoftmaxGrad)):
+        return
+
+    sm = None
+    try:
+        d_sm, sm = node.inputs
+    except Exception:
+        return
+
+    if not((sm is not None) and sm.owner and isinstance(sm.owner.op, Softmax)):
+        return
+
+    if d_sm.owner and isinstance(d_sm.owner.op, list_classes_IncSubtensor):
+        try:
+            incr = d_sm.owner.inputs[1]
+            subtensor_inputs, = d_sm.owner.inputs[2:]
+        except:
+            return
+
+        out_grad = 1.
+        if incr.owner and incr.owner.op == tensor.neg:
+            incr = incr.owner.inputs[0]
+            out_grad = - out_grad
+
+        if incr.owner and incr.owner.op == tensor.true_div:
+            num, denom = incr.owner.inputs
+
+            if num.ndim == 1 or np.all(num.broadcastable):
+                out_grad *= -num
+            else:
+                return
+
+            if not denom.owner:
+                return
+            if isinstance(denom.owner.op, list_classes_Subtensor):
+                # Base case
+                adv_subtensor = denom
+                # out_grad /= 1.
+            elif denom.owner.op == tensor.mul:
+                # Try to find the AdvancedSubtensor node mentionned above,
+                # and the output gradient
+                for i, input in enumerate(denom.owner.inputs):
+                    if input.owner and isinstance(input.owner.op, list_classes_Subtensor):
+                        other_inputs = [in_ for (j, in_) in enumerate(denom.owner.inputs) if j != i]
+                        if len(other_inputs) == 1:
+                            rest = other_inputs[0]
+                        else:
+                            rest = tensor.mul(*[other_inputs])
+                        # Check that rest is a vector or a scalar
+                        if rest.ndim == 1 or np.all(rest.broadcastable):
+                            adv_subtensor = input
+                            out_grad /= rest
+                            break
+            else:
+                return
+
+            if adv_subtensor is not None:
+                try:
+                    maybe_sm = adv_subtensor.owner.inputs[0]
+                    maybes_subtensor_inputs, = adv_subtensor.owner.inputs[1:]
```

Comment:
Actually, that only works for a subtensor with only one index. We want to make that work whatever the indices are.
Also, I'd name the variable `maybe_subtensor_idx` or something like that.